In [10]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
import json
import numpy as np

In [68]:
data = pd.read_csv('travel_data.csv')
data.head()

,id,city,country,region,short_description,latitude,longitude,avg_temp_monthly,ideal_durations,budget_level,culture,adventure,nature,beaches,nightlife,cuisine,wellness,urban,seclusion
0,c54acf38-3029-496b-8c7a-8343ad82785c,Milan,Italy,europe,"Chic streets lined with fashion boutiques, his...",45.464194,9.189635,"{""1"":{""avg"":3.7,""max"":7.8,""min"":0.4},""2"":{""avg...","[""Short trip"",""One week""]",Luxury,5,2,2,1,4,5,3,5,2
1,0bd12654-ed64-424e-a044-7bc574bcf078,Yasawa Islands,Fiji,oceania,"Crystal-clear waters, secluded beaches, and vi...",-17.290947,177.125786,"{""1"":{""avg"":28,""max"":30.8,""min"":25.8},""2"":{""av...","[""Long trip"",""One week""]",Luxury,2,4,5,5,2,3,4,1,5
2,73036cda-9134-46fc-a2c6-807782d59dfb,Whistler,Canada,north_america,Snow-capped peaks and lush forests create a se...,50.117190,-122.954302,"{""1"":{""avg"":-2.5,""max"":0.4,""min"":-5.5},""2"":{""a...","[""Short trip"",""Weekend"",""One week""]",Luxury,3,5,5,2,3,3,4,2,4
3,3872c9c0-6b6e-49e1-9743-f46bfe591b86,Guanajuato,Mexico,north_america,Winding cobblestone streets and colorful facad...,20.987700,-101.000000,"{""1"":{""avg"":15.5,""max"":22.8,""min"":8.7},""2"":{""a...","[""Weekend"",""One week"",""Short trip""]",Mid-range,5,3,3,1,3,4,3,4,2
4,e1ebc1b6-8798-422d-847a-22016faff3fd,Surabaya,Indonesia,asia,Bustling streets filled with the aroma of loca...,-7.245972,112.737827,"{""1"":{""avg"":28.1,""max"":32.5,""min"":25.5},""2"":{""...","[""Short trip"",""Weekend""]",Budget,4,3,3,2,3,4,3,4,2


In [70]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 560 entries, 0 to 559
Data columns (total 19 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 560 non-null    object 
 1   city               560 non-null    object 
 2   country            560 non-null    object 
 3   region             560 non-null    object 
 4   short_description  560 non-null    object 
 5   latitude           560 non-null    float64
 6   longitude          560 non-null    float64
 7   avg_temp_monthly   560 non-null    object 
 8   ideal_durations    560 non-null    object 
 9   budget_level       560 non-null    object 
 10  culture            560 non-null    int64  
 11  adventure          560 non-null    int64  
 12  nature             560 non-null    int64  
 13  beaches            560 non-null    int64  
 14  nightlife          560 non-null    int64  
 15  cuisine            560 non-null    int64  
 16  wellness           560 non

## Item-Item CF


In [84]:
# data cleaning
cf_data = data[['city', 'budget_level', 'culture', 'adventure', 'nature', 'beaches', 'nightlife', 'cuisine', 'wellness', 'urban', 'seclusion']]

cf_data.set_index('city', inplace=True)

cf_data.head()

,budget_level,culture,adventure,nature,beaches,nightlife,cuisine,wellness,urban,seclusion
city,,,,,,,,,,
Milan,Luxury,5,2,2,1,4,5,3,5,2
Yasawa Islands,Luxury,2,4,5,5,2,3,4,1,5
Whistler,Luxury,3,5,5,2,3,3,4,2,4
Guanajuato,Mid-range,5,3,3,1,3,4,3,4,2
Surabaya,Budget,4,3,3,2,3,4,3,4,2


In [90]:
onehot_X = pd.get_dummies(cf_data, drop_first=True, dtype = int)
onehot_X.head()

,culture,adventure,nature,beaches,nightlife,cuisine,wellness,urban,seclusion,budget_level_Luxury,budget_level_Mid-range
city,,,,,,,,,,,
Milan,5,2,2,1,4,5,3,5,2,1,0
Yasawa Islands,2,4,5,5,2,3,4,1,5,1,0
Whistler,3,5,5,2,3,3,4,2,4,1,0
Guanajuato,5,3,3,1,3,4,3,4,2,0,1
Surabaya,4,3,3,2,3,4,3,4,2,0,0


In [98]:
city_df = onehot_X.T
city_df

city,Milan,Yasawa Islands,Whistler,Guanajuato,Surabaya,Nuuk,Windhoek,Kingston,Nafplio,Tbilisi,...,Savusavu,Bristol,Rovaniemi,Viña del Mar,Nelson,Maun,Gothenburg,Manchester,Copenhagen,Sucre
culture,5,2,3,5,4,3,3,4,4,4,...,3,4,3,3,3,3,4,4,5,4
adventure,2,4,5,3,3,4,4,3,3,3,...,4,3,4,3,4,5,3,2,2,3
nature,2,5,5,3,3,5,4,3,4,3,...,5,3,5,4,5,5,4,2,3,3
beaches,1,5,2,1,2,2,1,3,3,1,...,5,2,1,4,3,1,3,1,2,1
nightlife,4,2,3,3,3,2,3,4,3,3,...,2,4,2,4,2,2,3,4,4,2
cuisine,5,3,3,4,4,3,3,4,4,4,...,3,4,3,4,3,3,4,4,4,3
wellness,3,4,4,3,3,2,3,3,3,3,...,4,3,3,3,3,3,3,3,3,2
urban,5,1,2,4,4,2,3,3,3,3,...,2,4,2,4,3,2,4,4,5,2
seclusion,2,5,4,2,2,5,4,2,4,3,...,5,2,4,2,4,4,3,2,2,4
budget_level_Luxury,1,1,1,0,0,1,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


In [ ]:
# Initial item-based collaborative filtering function
import numpy as np

def cosine_sim(x, y):
    return ((x.T @ y) /(np.sqrt(x.T @ x) * np.sqrt(y.T @ y))).item()

def min_max(a):
    a_min = a.min()
    a_max = a.max()

    min_max = (a - a_min) / (a_max - a_min)
    return min_max


def item_collab_filter(df, targ_user, sim_score='Cosine', k=2):
    # center data and get mean of each items
    centered_df = df.sub(df.mean(axis=0), axis=1)
    items_mean = df.mean(axis=0)

    missing_items = df.loc[targ_user, :][df.loc[targ_user, :].isna()].index

    pred_rating_dict = {}

    for item in missing_items:
        scores = {}
        for comp_item in centered_df.columns:
            # nan for same items in pairwise comparison
            if item == comp_item:
                scores[comp_item] = np.nan
            else:
                # Both items have to be rated
                pair = centered_df[[item, comp_item]].dropna()

                array1 = np.asarray(pair[item]).reshape(len(pair), 1)
                array2 = np.asarray(pair[comp_item]).reshape(len(pair), 1)

                # calculate similarity scores
                if sim_score == 'Cosine':
                    scores[comp_item] = cosine_sim(array1, array2)
                elif sim_score == 'L2':
                    scores[comp_item] = -np.linalg.norm(array1 - array2, ord=2)

        # min max scaling
        scores = pd.Series(scores)
        scores = min_max(scores)

        # find k most similar items
        k_similar_items = scores.dropna().nlargest(k)

        # fill in missing ratings with item means
        user_ratings = df.loc[targ_user, k_similar_items.index]
        user_ratings = user_ratings.fillna(items_mean)

        # predict rating
        pred_rating = sum(user_ratings * k_similar_items)/sum(k_similar_items)
        pred_rating_dict[item] = pred_rating

    return pred_rating_dict

## Continuation of Colloborative Filtering for POC
* For 560 destinations, the dataset contains rankings from one to five for a variety of aspects
* In order to conduct Item-Item Collaborative Filtering, we just took the rankings from the dataset, then made each city a column, and the ranking categories as the rows
* Thus, each category acts as a user and each city like an item in a traditional collaboritive filtering dataset
* We created an initial version of an item item collab filtering
* In order to test it, we will randomly remove ratings and then compare to actual ratings, and see how similar they are
* For real reccomendations, certain destinations would not have ratings for some categories, and if travelers want to know what rating a destination would have, they could use the collab filtering algorithm

### Data Exploration

In [35]:
data = pd.read_csv('travel_data.csv')
data.head()

,id,city,country,region,short_description,latitude,longitude,avg_temp_monthly,ideal_durations,budget_level,culture,adventure,nature,beaches,nightlife,cuisine,wellness,urban,seclusion
0,c54acf38-3029-496b-8c7a-8343ad82785c,Milan,Italy,europe,"Chic streets lined with fashion boutiques, his...",45.464194,9.189635,"{""1"":{""avg"":3.7,""max"":7.8,""min"":0.4},""2"":{""avg...","[""Short trip"",""One week""]",Luxury,5,2,2,1,4,5,3,5,2
1,0bd12654-ed64-424e-a044-7bc574bcf078,Yasawa Islands,Fiji,oceania,"Crystal-clear waters, secluded beaches, and vi...",-17.290947,177.125786,"{""1"":{""avg"":28,""max"":30.8,""min"":25.8},""2"":{""av...","[""Long trip"",""One week""]",Luxury,2,4,5,5,2,3,4,1,5
2,73036cda-9134-46fc-a2c6-807782d59dfb,Whistler,Canada,north_america,Snow-capped peaks and lush forests create a se...,50.117190,-122.954302,"{""1"":{""avg"":-2.5,""max"":0.4,""min"":-5.5},""2"":{""a...","[""Short trip"",""Weekend"",""One week""]",Luxury,3,5,5,2,3,3,4,2,4
3,3872c9c0-6b6e-49e1-9743-f46bfe591b86,Guanajuato,Mexico,north_america,Winding cobblestone streets and colorful facad...,20.987700,-101.000000,"{""1"":{""avg"":15.5,""max"":22.8,""min"":8.7},""2"":{""a...","[""Weekend"",""One week"",""Short trip""]",Mid-range,5,3,3,1,3,4,3,4,2
4,e1ebc1b6-8798-422d-847a-22016faff3fd,Surabaya,Indonesia,asia,Bustling streets filled with the aroma of loca...,-7.245972,112.737827,"{""1"":{""avg"":28.1,""max"":32.5,""min"":25.5},""2"":{""...","[""Short trip"",""Weekend""]",Budget,4,3,3,2,3,4,3,4,2


In [36]:
## Make dataframe for bayesian
data_bayes = pd.DataFrame()

# Get temperature features
temps = data['avg_temp_monthly'].apply(json.loads)

data_bayes['temp_annual_avg'] = temps.apply(lambda x: np.mean([x[m]["avg"] for m in x]))
data_bayes['temp_summer_avg'] = temps.apply(lambda x: np.mean([x[m]["avg"] for m in ['6','7','8']]))
data_bayes['temp_winter_avg'] = temps.apply(lambda x: np.mean([x[m]["avg"] for m in ['12','1','2']]))

# Duration separation
durations = data['ideal_durations'].apply(json.loads)

for label in ['Day trip', 'Weekend', 'Short trip', 'One week', 'Long trip']:
    col_name = f'duration_{label.lower().replace(" ","_")}'
    data_bayes[col_name] = durations.apply(lambda x: int(label in x))

# Rank budgets
data_bayes['budget_rank'] = data['budget_level'].map({
    'Budget': 1,
    'Mid-range': 2,
    'Luxury': 3
})

# Add latitude and longitude
data_bayes['latitude'] = data['latitude']
data_bayes['longitude'] = data['longitude']

# Get overall rating of destinations by averaging category ratings
data_bayes['overall_rating'] = data[['culture', 'adventure', 'nature', 'beaches', 'nightlife', 'cuisine', 'wellness', 'urban', 'seclusion']].mean(axis=1)

In [37]:
data_bayes.head()

,temp_annual_avg,temp_summer_avg,temp_winter_avg,duration_day_trip,duration_weekend,duration_short_trip,duration_one_week,duration_long_trip,budget_rank,latitude,longitude,overall_rating
0,14.750000,24.833333,5.166667,0,0,1,1,0,3,45.464194,9.189635,3.222222
1,26.975000,25.933333,28.066667,0,0,0,1,1,3,-17.290947,177.125786,3.444444
2,6.808333,16.700000,-2.033333,0,1,1,1,0,3,50.117190,-122.954302,3.444444
3,19.533333,21.166667,15.866667,0,1,1,1,0,2,20.987700,-101.000000,3.111111
4,28.758333,28.433333,28.200000,0,1,1,0,0,1,-7.245972,112.737827,3.111111


In [38]:
data_bayes.to_csv('travel_data_bayes.csv', index=False)